### PR TITLE
Issue 286 - expose object references in ReferenceChange

### PR DESCRIPTION
**Pull request by gtnarg**
Any comments or feedback welcomed :)

### CODE REVIEW

**Comment by bartoszwalacik** [comment]
The Travis CI build failed :(

File: javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy [hunk 1]
```diff
@@ -9,6 +9,8 @@ import spock.lang.Unroll
 import static java.util.Optional.*
```
Comment (mwesolowski) [inline javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy]: Javers can be run on JDK 7 and java.util.Optional was introduced in JDK 8. Please use our internal org.javers.common.collections.Optional instead

File: javers-core/src/main/java/org/javers/core/diff/appenders/ReferenceChangeAppender.java [hunk 2]
```diff
@@ -20,15 +20,16 @@ public boolean supports(JaversType propertyType) {
         return propertyType instanceof ManagedType;
     }
 
-    @Override
-    public ReferenceChange calculateChanges(NodePair pair, Property property) {
-        GlobalId leftId =  pair.getLeftGlobalId(property);
-        GlobalId rightId = pair.getRightGlobalId(property);
+	@Override
```
Comment (mwesolowski) [inline javers-core/src/main/java/org/javers/core/diff/appenders/ReferenceChangeAppender.java]: Please use 4 spaces indentation

File: javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy (hunk 1 above)
Comment (bartoszwalacik) [inline javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy]: @mwesolowski as a rule of thumb yes, you are right but there are exceptions
JaVers sits in the middle between Java7 and Java8

It's quite complicted, i have to clarify this in Contributor Guidlines ...
Whole /src should be written in Java7, except of packages like `package org.javers.java8support`
In /test Java8 is allowed

So in this case Java8 is OK

Big Picture is: we want JaVers to be runable on JDK7 but also want to support new Java type (`Java8AddOns`)

File: javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy [hunk 3]
```diff
@@ -50,9 +52,12 @@ class OptionalChangeAppenderTest extends AbstractDiffAppendersTest {
...
                     .hasInstanceId(SnapshotEntity, 1)
         change instanceof ReferenceChange
-        change.left == expectedLeftRef
-        change.right == expectedRightRef
-
+        ReferenceChangeAssert.assertThat( change )
+            .hasLeftReference( expectedLeftRef )
+            .hasRightReference( expectedRightRef )
+            .hasLeftObject( org.javers.common.collections.Optional.empty() )
+            .hasRightObject( org.javers.common.collections.Optional.empty() )
```
Comment (bartoszwalacik) [inline javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy]: why not
```
        change.leftObject == leftOptional
        change.rightObject == rightOptional
```
?

File: javers-core/src/test/groovy/org/javers/core/diff/appenders/ReferenceChangeAppenderTest.groovy [hunk 4]
```diff
@@ -37,13 +41,17 @@ class ReferenceChangeAppenderTest extends AbstractDiffAppendersTest {
         assertThat(change)
                   .hasLeftReference(SnapshotEntity,2)
                   .hasRightReference(null)
+                  .hasLeftObject(of(entityRef))
+                  .hasRightObject(empty())
```
Comment (bartoszwalacik) [inline javers-core/src/test/groovy/org/javers/core/diff/appenders/ReferenceChangeAppenderTest.groovy]: maybe:
```
.hasLeftObject(entityRef)
.hasNullRightObject()
```

File: javers-core/src/test/groovy/org/javers/core/diff/appenders/ReferenceChangeAssert.groovy [hunk 5]
```diff
@@ -32,13 +32,23 @@ class ReferenceChangeAssert {
...
     }
 
-    ReferenceChangeAssert hasRightReference(GlobalId expected) {
+    ReferenceChangeAssert hasRightReference(Object expected) {
         assert actual.right == expected
         this
     }
+    
+    ReferenceChangeAssert hasLeftObject(Object expected) {
+        assert actual.leftObject == expected
```
Comment (bartoszwalacik) [inline javers-core/src/test/groovy/org/javers/core/diff/appenders/ReferenceChangeAssert.groovy]: `assert actual.leftObject.get == expected`

File: javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy (hunk 3 above)
Comment (bartoszwalacik) [inline javers-core/src/test/groovy/org/javers/core/diff/appenders/OptionalChangeAppenderTest.groovy]: i mean why you expect that both leftObject and rightObjects are Optional.empty() ?

**Comment by gtnarg** [comment]
Sorry for the delay, was going to address your comments tonight...

**Pull request merged**